### PR TITLE
Support cron and timezone in manager task integration

### LINKED ISSUE #1739: Support `cron` in ScyllaCluster API for Scylla Manager tasks
### What should the feature do?

With https://github.com/scylladb/scylla-manager/pull/3720 and existing, partial deprecation of `start_date`, we should add support for [`cron`](https://github.com/scylladb/scylla-manager/blob/d86640b0c3f64599be6a56b4590f294a26bd07f2/swagger/scylla-manager.json#L270) in our API for Scylla Manager tasks. 

### What is the use case behind this feature?

ditto

### Anything else we need to know?

_No response_

### LINKED ISSUE #1796: Deprecate "interval" in ScyllaCluster's manager task API
### What should the feature do?

The `interval` parameter has been deprecated in the most recent Scylla Manager release - 3.2.6. We should deprecate it in our API as well.

### What is the use case behind this feature?

ditto

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by rzetelskik**
**Description of your changes:** This PR adds missing cron and related fields, currently missing from out API, to manager task specs and extends the unit tests to cover them.

**Which issue is resolved by this Pull Request:**
Resolves #1739 

/kind feature
/priority important-longterm
/cc

### CODE REVIEW

**Comment by rzetelskik** [comment]
/priority critical-urgent
since the bug fixes are a prerequisite for https://github.com/scylladb/scylla-operator/issues/1752

**Comment by rzetelskik** [comment]
/cc @Michal-Leszczynski 
would you be able to verify if there's anything missing from manager task specs in our API?

File: pkg/api/scylla/v1/types_cluster.go
Comment (tnozicka) [inline pkg/api/scylla/v1/types_cluster.go:326]: this is not in scope in the description of this PR, please split this so its visible in the release notes

File: pkg/api/scylla/v1/types_cluster.go
Comment (tnozicka) [inline pkg/api/scylla/v1/types_cluster.go:345]: `Therefore, we ask that pointers always be used with optional fields that do not have a built-in nil value.`
https://github.com/kubernetes/community/blob/e9f78a/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:326]: ok, https://github.com/scylladb/scylla-operator/pull/1853

**Review by tnozicka — CHANGES_REQUESTED** [review]
please fix the API to use pointers where needed

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:345]: > that do not have a built-in nil value

""?

Comment (tnozicka) [reply pkg/api/scylla/v1/types_cluster.go:345]: that's not `nil` value, but a `zero` value

Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:345]: for instance: https://github.com/scylladb/scylla-operator/blob/1520cf77e2bd089ad3497264dbd86196a4586f29/pkg/api/scylla/v1/types_cluster.go#L106

Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:345]: Ok, I see there are counterexamples in our api as well, so that's not really an argument. I'll amend it

Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go]: done

**Comment by Michal-Leszczynski** [comment]
Feel free to ignore my comments if they are not related to this PR. I just wanted to highlight some fields that perhaps require additional attention and give context for some of them.

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:403]: > This is not repair-specific field

The api on our side doesn't really match SM's, so there isn't really a better place for it - the shared inline struct is `SchedulerTaskSpec`, which doesn't fit the case here. The fields in RepairTaskSpec/BackupTaskSpec seem to be a mix of mapped properties and some shared fields, some even taken out of the `Schedule`on SM side. tl;dr; it's a mess that we have to live with now for backwards compatibility

> Do you have a use case for this field?

Nope, I just didn't want to leave it out while synchronising all the other fields.

> first validate its existence on SM side.

Well, it's in the model https://github.com/scylladb/scylla-manager/commit/5a73edfd5b2c5b344829ddd58adec641d965d245

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:333]: It's an api convention, @tnozicka even pointed it out higher up in the comments https://github.com/scylladb/scylla-operator/pull/1851#discussion_r1531841953.
This particular field has a default, so it doesn't exactly make sense. Out of scope for this PR anyway.

**Comment by rzetelskik** [comment]
> The now syntax can cause a lot of pain since it is parsed on the operator or sctool side. Maybe it's worth to discourage users from using via appropriate field description?

Agreed. Is that mentioned in sctool/sm? 

> The interval field is (or should be) deprecated by SM. Is it also deprecated here?

There's a separate PR in queue with interval deprecation. Best we can do is a field description though.

> Why is it string? SM supports float64 for backward compatibility, but all floats are converted to ints anyway, because they don't make sense anymore from Scylla POV.

To be honest - no idea. I don't think we can change this in this API version though.

File: pkg/api/scylla/v1/types_cluster.go
Comment (tnozicka) [reply pkg/api/scylla/v1/types_cluster.go]: I agree mistakes were made :)

thanks for the update

File: pkg/api/scylla/v1/types_cluster.go
Comment (Michal-Leszczynski) [reply pkg/api/scylla/v1/types_cluster.go:439]: But then there is a difference between operator and SM.
Running `sctool backup --retention-days 3` and corresponding (in a direct way) operator definition results in a different retention policy. Of course it's up to operator team to decide that, but I just wanted to highlight that.

File: pkg/api/scylla/v1/types_cluster.go
Comment (tnozicka) [reply pkg/api/scylla/v1/types_cluster.go:403]: I would scope this down to the fields we see a meaning for. We can always add a field but never to remove one. So let's avoid adding them just because they are present in manager API. I don't see a big reason for `IgnoreDownHosts` either at this point.

Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go]: Removed the tags. As for `IgnoreDownHosts`, I think a reasonable rule of thumb is to allow setting whatever sctool allows for.

**Comment by rzetelskik** [comment]
split mapstructure fix into a separate PR: https://github.com/scylladb/scylla-operator/pull/1854
/hold for https://github.com/scylladb/scylla-operator/pull/1854

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go]: @tnozicka after internal conversation I removed the other fields, only leaving `cron` and `timezone` related to it

**Comment by rzetelskik** [comment]
ping @tnozicka

File: pkg/api/scylla/v1/types_cluster.go
Comment (tnozicka) [reply pkg/api/scylla/v1/types_cluster.go]: > imo the only reasonable thing to do would be to drop the defaults on our side altogether

you can stop defaulting the field but you can't make it required

We should not use defaults, if we want the manager to accept optional values. But that ship has half sailed. We default at admission and after that point the values are stored and always used as explicitly specified.

And defaulting or the lack of shall not render object invalid.

Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go]: I'll verify which defaults could be dropped safely, but that's for a separate PR anyway, since this ended up just adding cron and timezone. So let's get back to it once all the fixes in queue land, I don't want to introduce extra variables.

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go]: removed

File: pkg/api/scylla/v1/types_cluster.go
Comment (tnozicka) [inline pkg/api/scylla/v1/types_cluster.go:341]: if this is truly only for `cron`, maybe it should be named `cronTimezone` to avoid implications on `startDate` field?

Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:341]: It's also for 'window', which was excluded from this PR. I think naming it that way would kill potential extensibility. And I would refrain from interfering with the upstream api to much in general, it makes the ux worse imo.

Comment (tnozicka) [reply pkg/api/scylla/v1/types_cluster.go:341]: > It's also for 'window',

ok, with that is makes sense.

Comment (tnozicka) [inline pkg/api/scylla/v1/types_cluster.go:341]: please add validation making sure it's set only if the fields that it applies to are present

File: pkg/api/scylla/v1/types_cluster.go
Comment (tnozicka) [inline pkg/api/scylla/v1/types_cluster.go:337]: is this mutually exclusive with start date? should you validate both are not set?

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:341]: wouldn't it be better to make it transparent, let the client return an error and propagate it to status (globally)?

File: pkg/controller/manager/types_old.go
Comment (zimnx) [inline pkg/controller/manager/types_old.go:43]: is it part of webhook validation? If not, please move it there.

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:337]: I don't see sctool doing this. Looking at https://github.com/scylladb/scylla-manager/pull/3720#issuecomment-1956621668 and https://github.com/scylladb/scylla-manager/issues/3750 I'm not even sure I understand how they interact with each other. @Michal-Leszczynski can you advise?

File: pkg/controller/manager/types_old.go
Comment (rzetelskik) [reply pkg/controller/manager/types_old.go:43]: I'll revert this change here as it's unrelated and I'll open a separate PR for it. Preferably after https://github.com/scylladb/scylla-operator/pull/1853 and https://github.com/scylladb/scylla-operator/pull/1850 land.

Comment (rzetelskik) [reply pkg/controller/manager/types_old.go]: reverted

File: pkg/api/scylla/v1/types_cluster.go
Comment (Michal-Leszczynski) [reply pkg/api/scylla/v1/types_cluster.go:337]: @karol-kokoszka please correct me if I'm wrong, but `cron` works with `startDate` in a way that the first `cron` activation will scheduled after the `startDate` (not necessarily on the `startDate` itself). So the `startDate` can be perceived as a date when the `cron` task is added to SM.

Comment (karol-kokoszka) [reply pkg/api/scylla/v1/types_cluster.go:337]: > but cron works with startDate in a way that the first cron activation will scheduled after the startDate

Indeed.

Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:337]: so to my understanding cron is mutually exclusive with interval, but both can work with start date?

Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:337]: Added validation for mutual exclusivity of cron and interval. A bit awkward since interval is not a pointer and defaulted.

Comment (tnozicka) [reply pkg/api/scylla/v1/types_cluster.go:336]: yeah, I think even with the startDate it's weird if your `startDate` is say on Monday and cron says it should run on Fridays. Lets keep the validation tight as we can never go back, but loosening is possible if a use case is found that is more valuable then the conflict I pointed out here.

File: pkg/api/scylla/v1/types_cluster.go
Comment (tnozicka) [reply pkg/api/scylla/v1/types_cluster.go:340]: >  the timezone will just be ignored

well, that suck even a bit more, since the user has a timezone set thinking all is well, but it silently runs in a different time then he though

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:336]: I'll add the validation here, also sent a separate PR moving all existing validation to the validating webhook: https://github.com/scylladb/scylla-operator/pull/1868.
I also created a tracker for the other fields which won't be cover in these two PRs: https://github.com/scylladb/scylla-operator/issues/1869.

File: pkg/api/scylla/validation/cluster_validation.go
Comment (zimnx) [inline pkg/api/scylla/validation/cluster_validation.go:245]: duplicated "invalid"

File: pkg/api/scylla/validation/cluster_validation.go
Comment (rzetelskik) [reply pkg/api/scylla/validation/cluster_validation.go:224]: that'll be a part of https://github.com/scylladb/scylla-operator/pull/1868, already removed there

File: pkg/api/scylla/validation/cluster_validation.go
Comment (rzetelskik) [reply pkg/api/scylla/validation/cluster_validation.go:245]: that'll be a part of https://github.com/scylladb/scylla-operator/pull/1868, already removed there

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:335]: done

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:339]: done

**Comment by rzetelskik** [comment]
@zimnx @tnozicka I believe I have addressed all your comments - please have another go.

File: pkg/api/scylla/v1/types_cluster.go
Comment (rzetelskik) [reply pkg/api/scylla/v1/types_cluster.go:335]: @tnozicka I wonder, should we also validate the extended cron format on our side? We don't control it, so if it changes upstream, we'll get out of sync

Comment (tnozicka) [reply pkg/api/scylla/v1/types_cluster.go:335]: I'd start with a strict validation of the format and loosen if we bump into something. Kube seems to use a cron library as well, so this might be a good start https://github.com/kubernetes/kubernetes/blob/e800d15/go.mod#L58C2-L58C27

File: pkg/api/scylla/validation/cluster_validation.go
Comment (tnozicka) [inline pkg/api/scylla/validation/cluster_validation.go:255]: please add timezone and cron format validation

Comment (rzetelskik) [reply pkg/api/scylla/validation/cluster_validation.go:255]: @tnozicka Scylla Manager accepts the `timezone` parameter, which, at least according to sctool docs, represents `Timezone of –cron and –window flag values`.

At the same time, it accepts `TZ` and `CRON_TZ` prefixes of `cron` parameter (see https://github.com/robfig/cron/blob/e843a09e5b2db454d77aad25b1660173445fb2fc/parser.go#L95).

The combination of two is accepted. K8s has a similar issue with CronJobs, and it doesn't accept TZ/CRON_TZ prefixes: see https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#unsupported-timezone-specification.

Should we handle it the same way k8s does? Should I raise this with the manager team?

Comment (tnozicka) [reply pkg/api/scylla/validation/cluster_validation.go:255]: I think so + stricter is always better start with the validation

Comment (rzetelskik) [reply pkg/api/scylla/validation/cluster_validation.go:278]: ok, done

**Comment by rzetelskik** [comment]
@tnozicka I added cron and timezone validation. TZ and CRON_TZ are not accepted as part of cron for now. I'll raise an issue with the manager team.

**Comment by rzetelskik** [comment]
manager flake and upgrade flake - can't possibly be deteriorated by this PR
https://github.com/scylladb/scylla-operator/issues/1694#issuecomment-2066105109
https://github.com/scylladb/scylla-operator/issues/1233#issuecomment-2066106629

/test images
/retest

File: pkg/api/scylla/validation/cluster_validation.go
Comment (tnozicka) [inline pkg/api/scylla/validation/cluster_validation.go:268]: nit: please split the assignment to make it shorter

File: pkg/api/scylla/validation/cluster_validation_test.go
Comment (tnozicka) [inline pkg/api/scylla/validation/cluster_validation_test.go:268]: it's unclear whether the test should produce and error or not from it's name (in other places as well), especially given it says `invalid` and produces no error

**Review by tnozicka — COMMENTED** [review]
thanks, the api and overall logic looks sound to me, had some technical comments

File: pkg/api/scylla/validation/cluster_validation_test.go
Comment (rzetelskik) [reply pkg/api/scylla/validation/cluster_validation_test.go:268]: my bad, it shouldn't say invalid here

Comment (rzetelskik) [reply pkg/api/scylla/validation/cluster_validation_test.go]: done

File: pkg/api/scylla/validation/cluster_validation.go
Comment (rzetelskik) [reply pkg/api/scylla/validation/cluster_validation.go]: done

**Comment by rzetelskik** [comment]
> @rzetelskik: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> ci/prow/e2e-gke-release-script-latest	[415c60b](https://github.com/scylladb/scylla-operator/commit/415c60b1b0dcf5fe7a85ec399e8ce47c4fd79bf2)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/1851/pull-scylla-operator-master-e2e-gke-release-script-latest/1773008427048505344)	true	`/test e2e-gke-release-script-latest`
> ci/prow/e2e-gke-parallel-clusterip	[7ea90ba](https://github.com/scylladb/scylla-operator/commit/7ea90bae594da3591cffac1567476f6493c50ff2)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/1851/pull-scylla-operator-master-e2e-gke-parallel-clusterip/1781267441016377344)	true	`/test e2e-gke-parallel-clusterip`
> ci/prow/e2e-gke-parallel	[7ea90ba](https://github.com/scylladb/scylla-operator/commit/7ea90bae594da3591cffac1567476f6493c50ff2)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/1851/pull-scylla-operator-master-e2e-gke-parallel/1781267440487895040)	true	`/test e2e-gke-parallel`
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=1851). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).

manager flake
/retest

**Comment by rzetelskik** [comment]
same thing again
/retest

**Comment by rzetelskik** [comment]
and again
/retest

**Review by tnozicka — APPROVED** [review]
thanks for the updates

/approve
high level lgtm but I only scanned the test
/assign @zimnx

**Comment by rzetelskik** [comment]
aaand again

/test images
/retest

File: pkg/api/scylla/validation/cluster_validation.go
Comment (rzetelskik) [reply pkg/api/scylla/validation/cluster_validation.go:279]: Xref https://github.com/scylladb/scylla-manager/issues/3818

**Review by zimnx — APPROVED** [review]
thanks
/lgtm

**Pull request merged**
